Fix Iceberg partition tables

linxingyuan1102 opened the pull request:
This commit includes two fixes:
1. Use a table scan which loads column stats with each data file.
   Otherwise the column stats returned by Iceberg are null.
2. Make minValues and maxValues in class Partition mutable so that
   Partition::updateStats doesn't throw an error.

cc: @electrum @Parth-Brahmbhatt

sopel39 commented on presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/PartitionTable.java:
```diff
@@ -339,8 +339,8 @@ public Partition(
-            this.minValues = ImmutableMap.copyOf(requireNonNull(minValues, "minValues is null"));
-            this.maxValues = ImmutableMap.copyOf(requireNonNull(maxValues, "maxValues is null"));
+            this.minValues = new HashMap<>(requireNonNull(minValues, "minValues is null"));
+            this.maxValues = new HashMap<>(requireNonNull(maxValues, "maxValues is null"));
```
shouldn't update stats produce new `Partition` object or at least `minValues/maxValues` maps.
Is `Partition` used by multiple threads?

electrum commented on presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/PartitionTable.java:
```diff
@@ -339,8 +346,8 @@ public Partition(
             this.size = size;
-            this.minValues = ImmutableMap.copyOf(requireNonNull(minValues, "minValues is null"));
-            this.maxValues = ImmutableMap.copyOf(requireNonNull(maxValues, "maxValues is null"));
+            this.minValues = new HashMap<>(requireNonNull(minValues, "minValues is null"));
```
Wrap the copy so that the returned references cannot be modified:
```java
this.minValues = unmodifiableMap(new HashMap<>(minValues));
```

linxingyuan1102 replied:
But in this way `updateStats` will not be able to modify them neither? Did you mean to wrap them in these getter methods?

electrum replied:
Ah, my mistake, I didn't realize this is a mutable class. I thought we were changing from `ImmutableMap` due to having null values in the map. (which I missed in the commit message)

So ignore this comment. If this was a public class, we'd probably want to return a copy in the getter, but this isn't needed since this is a private helper class.

linxingyuan1102 commented:
Umbrella issue: #1324

electrum commented on presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/IcebergMetadata.java:
```diff
@@ -152,7 +153,7 @@ public IcebergTableHandle getTableHandle(ConnectorSession session, SchemaTableNa
             Configuration configuration = getConfiguration(session, tableName.getSchemaName());
             org.apache.iceberg.Table icebergTable = getIcebergTable(table.getSchemaName(), table.getTableName(), configuration, metastore);
-            return Optional.of(new PartitionTable(table, session, typeManager, icebergTable));
+            return Optional.of(new ClassLoaderSafeSystemTable(new PartitionTable(table, session, typeManager, icebergTable), getClass().getClassLoader()));
```
Maybe split this up:
```java
SystemTable systemTable = new PartitionTable(table, session, typeManager, icebergTable);
return Optional.of(new ClassLoaderSafeSystemTable(systemTable, getClass().getClassLoader()));

electrum commented on presto-spi/src/main/java/io/prestosql/spi/connector/classloader/ClassLoaderSafeSystemTable.java:
```diff
@@ -0,0 +1,70 @@
+
+import static java.util.Objects.requireNonNull;
+
+public class ClassLoaderSafeSystemTable
```
Add this to `TestClassLoaderSafeWrappers`

electrum commented:
The code is simple and looks good, but it's hard to tell what was broken/fixed. Can you add a test that fails without this change?

linxingyuan1102 commented:
> The code is simple and looks good, but it's hard to tell what was broken/fixed. Can you add a test that fails without this change? An integration test in `TestIcebergSmoke` is probably best.

Sure. I've added such a test in `TestIcebergSmoke`. I also have to add the support for `RowType` conversion in `TestingPrestoClient` so that the test case can query the partitions table which has row type columns.

linxingyuan1102 commented:
There were 2 check-style violations before so I updated the PR. I also rebased it upon the latest master so that a failure in Travis is fixed. Now the Travis passes.

electrum commented:
Merged, thanks!
